geiger/src/util/Collection.js

```diff
@@ -0,0 +1,144 @@
+import {isArray} from 'util';
+import defaults from 'lodash.defaults';
+import {isObject} from './util';
+
+/**
+ * @typedef {Object} CollectionOption
+ * @property {boolean} trackEnter - If true, Collection will track
+ *     newly added items. Need to manually clear.
+ * @property {boolean} trackExit - If true, Collection will track
+ *     newly removed items. Need to manually clear.
+ */
+
+/**
+ * A module that supports fast add/remove/get by id.
+ * So far, it is just a dictionary, may be updated later.
+ * Optionally, it can track the newly added and removed items.
+ *
+ * @class Collection
+ */
+export default class Collection {
+  /**
+   * @constructor
+   * @param {?Array<T>} items - Array of items
+   * @param {?CollectionOption} option
+   */
+  constructor(items, option) {
+    /**
+     * A mapping from event name to a list of event handlers.
+     * @type {!Object<string, T>}
+     * @private
+     */
+    this._items = {};
+
+    option = defaults(option, {trackEnter: false, trackExit: false});
+
+    this._trackEnter = option.trackEnter;
+    this._trackExit = option.trackExit;
+
+    if (option.trackEnter) {
+      /**
+       * Store all newly added items, reset by `clearEnterExit`
+       * @type {!Array<T>}
+       * @private
+       */
+      this._enter = [];
+    }
+
+    if (option.trackExit) {
+      /**
+       * Store all removed items, reset by `clearEnterExit`
+       * @type {!Array<T>}
+       * @private
+       */
+      this._exit = [];
+    }
+
+    if (items && isArray(items)) {
+      this.addItems(items);
+    }
+  }
+
+  add(id, item) {
+    if (arguments.length !== 2) {
+      throw new Error('Should have two arguments, but got: ' +
+          arguments.length);
+    }
+    if (typeof id === 'undefined') {
+      throw new Error('id should not be undefined');
+    }
+    if (this._trackEnter) {
+      this._enter.push(item);
+    }
+    this.set(id, item);
+    return this;
+  }
+
+  /**
+   * Add an array of items, each item should have id.
+   * @param {Array<T>} items
+   */
+  addItems(items) {
+    items.forEach((item) => {
+      if (isObject(item)) {
```

Comment:
Perhaps an assert on `id` is required?

Reply:
If an item has no `id`,  an error will be thrown inside `addItem()`. That's why I didn't add an assert here. But I can add here.